src/udata.rs

```diff
@@ -1,97 +1,408 @@
+        // returns a new placeholder type such that is_placeholder returns true
+        fn placeholder() -> Self {
+            PoseidonHash::Placeholder
+        }
+
+        // returns an empty hash such that is_empty returns true
+        fn empty() -> Self {
+            PoseidonHash::Empty
+        }
+
+        // returns true if this is a placeholder. This should be true iff this type was created by
+        // calling placeholder.
+        fn is_placeholder(&self) -> bool {
+            matches!(self, PoseidonHash::Placeholder)
+        }
+
+        // returns true if this is an empty hash. This should be true iff this type was created by
+        // calling empty.
+        fn is_empty(&self) -> bool {
+            matches!(self, PoseidonHash::Empty)
+        }
+
+        // used for serialization, writes the hash to the writer
+        //
+        // if you don't want to use serialization, you can just return an error here.
+        fn write<W>(&self, writer: &mut W) -> std::io::Result<()>
+        where
+            W: std::io::Write,
+        {
+            match self {
+                PoseidonHash::Hash(h) => writer.write_all(&h.to_bytes_be()),
+                PoseidonHash::Placeholder => writer.write_all(&[0u8; 32]),
+                PoseidonHash::Empty => writer.write_all(&[0u8; 32]),
+            }
+        }
+
+        // used for deserialization, reads the hash from the reader
+        //
+        // if you don't want to use serialization, you can just return an error here.
+        fn read<R>(reader: &mut R) -> std::io::Result<Self>
+        where
+            R: std::io::Read,
+        {
+            let mut buf = [0u8; 32];
+            reader.read_exact(&mut buf)?;
+            if buf.iter().all(|&b| b == 0) {
+                Ok(PoseidonHash::Empty)
+            } else {
+                Ok(PoseidonHash::Hash(Felt::from_bytes_be(&buf)))
+            }
+        }
+
+        // the main thing about the hash type, it returns the next node's hash, given it's children.
+        // The implementation of this method is highly consensus critical, so everywhere should use the
+        // exact same algorithm to calculate the next hash. Rustreexo won't call this method, unless
+        // **both** children are not empty.
+        fn parent_hash(left: &Self, right: &Self) -> Self {
+            if let (PoseidonHash::Hash(left), PoseidonHash::Hash(right)) = (left, right) {
+                return PoseidonHash::Hash(poseidon_hash_many(&[*left, *right]));
+            }
+
+            // This should never happen, since rustreexo won't call this method unless both children
+            // are not empty.
+            unreachable!()
+        }
+    }
+
+    impl Serialize for PoseidonHash {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            match self {
+                PoseidonHash::Hash(h) => {
+                    let inner = h.to_raw();
```

Comment:
Raw is the Montgomery form which is not very convenient to work with, I'd suggest to use `to_fixed_hex_string` instead